### ISSUE TITLE
Erroneous data in Home Assistant, or data in completely the wrong place

### ISSUE THREAD

**Issue by bodger-uk** [dan-r/HomeAssistant-Ohme#98]
I charge my car at 23:30 (due to cheap electricity from Octopus), and it charges until 03:30 - 05:00 depending on amount of charge needed. The data from Octopus shows this usage.

The data from Ohme may show the charge all took place within one hour at 09:00! Or, like last night, might show the charge correctly, but then give me an erroneous entry at 08:00. This makes my daily chart incorrect, and the individual devices part has an incorrect number.

This happens every time. With the first scenario I thought maybe the connection wasn't working and it was picking it all up on the connection refreshing. But with the second scenario I'm at a loss. It's reporting something that didn't happen.

Here's some pictures:

![Screenshot_20241007-104830](https://github.com/user-attachments/assets/096c70ff-6a37-463d-b3f7-dda16d725f26)
This was last night's charge

![Screenshot_20241007-104943](https://github.com/user-attachments/assets/963fd4cc-f447-4c83-b614-21ac3e01f113)
And this is what it looks like in HA

![Screenshot_20241007-104917](https://github.com/user-attachments/assets/734c259b-5bd4-48c8-b034-b924d604d5f8)
This was on Friday

![Screenshot_20241007-105001](https://github.com/user-attachments/assets/e84ed1bf-d449-4b3b-adb2-e57a46b81376)
And this was HA. This one has the Octopus data too.

Any help appreciated. This is really messing up my energy dashboard.

**Comment by dan-r** [comment]
What do your home assistant logs show? 

For my charge last night the energy sensor and Ohme app line up, so my money is on a comms issue, either between Home Assistant and Ohme or between your Ohme charger and Ohme.

**Comment by bodger-uk** [comment]
The only Ohme error in HA is:

This error originated from a custom integration.

Logger: custom_components.ohme.coordinator
Source: helpers/update_coordinator.py:386
integration: Ohme (documentation, issues)
First occurred: 5 October 2024 at 12:03:51 (3 occurrences)
Last logged: 6 October 2024 at 23:53:45

Error fetching Ohme Advanced Settings data: Error communicating with API

The last logged time is not long after I plugged the car in. Nothing in the logs for today.

**Comment by dan-r** [comment]
What about the Ohme 'Connected' sensor in Home Assistant? The app may play 'catch-up' if the charger is offline but the Home Assistant integration relies on the energy data coming in close to 'real time'.

**Comment by bodger-uk** [comment]
Is that the "charger online" sensor?

![image](https://github.com/user-attachments/assets/0405d4d8-4f0f-42f9-9491-f7f696b022b0)

**Comment by bodger-uk** [comment]
Did another charge last night. Started at 00:33, finished at 05:36:

![image](https://github.com/user-attachments/assets/c577fc2d-298d-48f0-97be-b95828dd2702)

The current Draw is correct:

![image](https://github.com/user-attachments/assets/60b079d8-e4c2-49b4-92fd-66961dcf9a71)

But the session energy is completely wrong:

![image](https://github.com/user-attachments/assets/8bf213b4-7e4d-45ba-bfea-cfbf80eff388)

and it's the session energy sensor.ohme_session_energy which is used by my energy dashboard. As a result, my energy usage is in the wrong places: 

![image](https://github.com/user-attachments/assets/ff2c870a-115a-4f47-b88e-c0ba4c0c7174)

and shows the wrong amount:

![image](https://github.com/user-attachments/assets/14e1d1aa-4b0b-4604-b4c7-3052a514f385)

The power draw reported by Ohme is a solid 7.5KW for the charge period

![image](https://github.com/user-attachments/assets/0211f5f9-d60a-4933-ba72-2703f63fc1e1)

I don't know what, if anything, I'm doing wrong. But all the sensors apart from the one the energy dashboard uses seem correct. How can I get sensor.ohme_session_energy to report correctly?

**Comment by dan-r** [comment]
What version of the integration are you running?

**Comment by bodger-uk** [comment]
1.0.0

**Comment by townsmcp** [comment]
+1 for having the same issues and errors in HA log also on 1.0.0

**Comment by BenCathro** [comment]
The sensor you're using to track the session energy appears to be cumulative. The charging session begins with the sensor already at about 30kwh.

**Comment by BenCathro** [comment]
I can't speak to the times being wrong as I'm still quite new to Home assistant and this integration. All I know is that my sensor of the same name resets to 0 every time I plug the car in.

Hopefully Dan is able to help. In the meantime have you tried to remove the integration from home assistant and configure again from scratch?

![image](https://github.com/user-attachments/assets/64f66012-1854-4744-a6e2-f6f3f20dec6c)

**Comment by bodger-uk** [comment]
This is the entire track for that sensor

![image](https://github.com/user-attachments/assets/947e8fc6-1bc9-49ab-934a-014691fa3b09)

I can certainly try removing it and reinstalling before I next charge the car (probably Thursday night)

**Comment by dan-r** [comment]
So this behaviour may be related to some logic around the energy sensor. Basically readings are ignored if they go backwards unless they go to 0, because for some reason the figures from Ohme sometimes go down then up again.

This seemed to really upset the Home Assistant energy dashboard and gave some negative readings, hence the logic.

I've added some debug logging to the relevant section, would you mind doing the following please:
1. Update to [v1.0.1-beta.1](https://github.com/dan-r/HomeAssistant-Ohme/releases/tag/v1.0.1-beta.1)
2. Enable debug logging for the Ohme integration
![image](https://github.com/user-attachments/assets/ecdf2cb5-e38b-4b8a-b2b5-19984219d688)

4. Provide the logs here. I only really need the lines containing `EnergyUsageSensor`, but can filter them.

**Comment by bodger-uk** [comment]
I updated to 1.0.1 as requested on Thursday and charged the car on Thursday night / Friday morning. As per the bug above, the used energy was not reported by Ohme during the charge. All of the charge utilisation was triggered when I unplugged the charger later on Friday morning

![image](https://github.com/user-attachments/assets/760278e8-38e9-464e-886a-f2befee03436)

I had debugging enabled, and filtered the log output to bring up EnergyUsageSensor. The filtered log is attached

[Ohme Log.txt](https://github.com/user-attachments/files/17532060/Ohme.Log.txt)

I appreciate your efforts Dan! Many thanks

**Comment by dan-r** [comment]
Thanks for sending that through, looks like it is happening for the reason I thought it was. The Wh value never resets to 0, in your case it drops to 18 and the integration subsequently ignores all the readings - I can fix this.

```
2024-10-24 22:28:18.134 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: Raw Wh reading 38617
2024-10-24 22:28:18.134 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: New state is 38617
2024-10-24 22:28:48.171 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: Raw Wh reading 38617
2024-10-24 22:28:48.172 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: New state is 38617
2024-10-24 22:29:17.929 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: Raw Wh reading 18
2024-10-24 22:29:17.929 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: New state is 38617
2024-10-24 22:29:48.491 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: Raw Wh reading 60
2024-10-24 22:29:48.491 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: New state is 38617
2024-10-24 22:30:17.921 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: Raw Wh reading 117
2024-10-24 22:30:17.921 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: New state is 38617
```

But whats really strange is that at 8am, the Wh value jumps to 66000. Was this when you unplugged?

```
2024-10-25 08:00:20.923 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: Raw Wh reading 16899
2024-10-25 08:00:20.924 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: New state is 38617
2024-10-25 08:00:51.011 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: Raw Wh reading 16957
2024-10-25 08:00:51.011 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: New state is 38617
2024-10-25 08:01:20.927 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: Raw Wh reading 17014
2024-10-25 08:01:20.928 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: New state is 38617
2024-10-25 08:01:51.172 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: Raw Wh reading 66025
2024-10-25 08:01:51.172 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: New state is 66025
2024-10-25 08:02:21.087 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: Raw Wh reading 66025
2024-10-25 08:02:21.087 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: New state is 66025
```

**Comment by bodger-uk** [comment]
That's it exactly. The second I unplugged the car I got a spike for all(?) the energy the charge had taken.

**Comment by dan-r** [comment]
How big is your battery/how much would you expect to have delivered in the charge?

**Comment by dan-r** [comment]
So there's actually 4 different watt-hour readings that come back from the chargeSessions API call. I've added debug logging in v1.0.1-beta.3 to print out all of these. Would you mind testing this version out on your next charge with debug logging on? I only need the lines that look like this:

`2024-10-27 19:49:28.657 DEBUG (MainThread) [custom_components.ohme.sensor] EnergyUsageSensor: CG: 919.0, BS: 909, CB: 929, CV: 909`

I need to charge tonight so will also test it out. For the record, I don't have my car 'API connected' to Ohme, so that might be why its working differently for you (assuming yours is). Of course the aim is to make it work equally for either route!

Thanks for your patience with this, its one of the many joys of working with undocumented APIs :)

**Comment by bodger-uk** [comment]
My next charge is Tuesday night / Wed morning. I'll have beta 3 in place before that.

I checked the charge, and the Ohme service in HA reported 27.41KW at the point I unplugged it, but the charge was only 22.5KW. I've seen discrepancies like that on other charges too.

**Comment by dan-r** [comment]
@bodger-uk thank you for sending the logs through and apologies for the delay - I haven't had much of a chance to work on this.

I've changed the source of the data used for the energy sensor in v1.0.1-beta.4, would you mind testing this out?

**Comment by bodger-uk** [comment]
Sure, I will be charging the car tonight or tomorrow night, so I'll have everything in place by then.

Thanks, again, for all your help on this.

**Comment by bodger-uk** [comment]
I charged the car last night, and put Ohme 1.0.1-beta4 into debug mode a few seconds before I plugged it in. I disconnected this morning, and closed the log. I lost my internet and some of my home network for around 25 minutes not long after I plugged the car in, so the log may have a gap between 00:00 and 00:30.

[ohme_log-2024-11-12.log](https://github.com/user-attachments/files/17714077/ohme_log-2024-11-12.log)

I've attached the whole thing this time because I can't see anything of use in it.

**Comment by dan-r** [comment]
Right so I've released v1.0.1-beta.5, which seems to work well for me with the energy dashboard. Would you mind testing it with your next charge?

**Comment by bodger-uk** [comment]
Sure, no problems. It'll be Thursday or Saturday night depending on charge levels, but I'll make sure to grab the logs when it charges

**Comment by bodger-uk** [comment]
I charged the car last night, and the entries are in the correct place and have the correct values on the dashboard. I do have the log, are there any particular lines you'd like?

Initial thoughts (after 1 data point!): much improved. Thanks for working this through. I'm likely to do another charge on Sunday night, so I can check again then (and log if required)

**Comment by dan-r** [comment]
Nothing I need to see from the logs - I'm just keeping an eye on the numbers over a few charges and it'd be great if you could do the same!

**Comment by bodger-uk** [comment]
A (maybe) final update. I have now charged the car 3 times using 1.0.1-beta 5 twice, and the official 1.0.1 once. In all three charges the data is correct and in the correct place. Matches up nicely with my Octopus data.

Thanks, again, for all your efforts on this, it's really appreciated.

I'm happy to mark this as closed if you are.

One little thing, the integration still says 1.0.0
![image](https://github.com/user-attachments/assets/72a1483a-f647-49a0-a332-0f9a264b6f6c)
definitely not important!

**Comment by dan-r** [comment]
That's good to hear! I'll keep an eye on it for my next charge or two then release it.

The version number has never actually been changed from 1.0.0, (even though I started at 0.1.0) so more of a 'a broken clock is right twice in a day' thing really :) I should start changing it though, as it would make troubleshooting a bit easier...

**Comment by dan-r** [comment]
Closing this issue as I've just released 1.1.0 with the new logic in. Please re-open this or open a new issue if you have any more issues with the energy data.

Dan
